### PR TITLE
Template lambda for scalar calls

### PR DESCRIPTION
**Pull request by Yurlungur**
## PR Summary

The work @gopsub is doing requires passing in many mesh-sized arrays into the lambda parameter. Currently on a per-cell basis, this requires packing those arrays into local contiguous arrays for lambdas. This can be cumbersome and expensive, especially for EOSs where that data is unneeded and is just a no-op.

This PR adds indexers/accessors to the lambda parameter for scalar calls, meaning data can now be passed in in a layout agnostic way. This should also reduce pain points in phoebus/riot, where the lambda arrays are used regularly and currently require contiguous accesses.

## PR Checklist

- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by using the `make format` command after configuring with `cmake`.
- [x] Document any new features, update documentation for changes made.
- [x] Make sure the copyright notice on any files you modified is up to date.
- [x] After creating a pull request, note it in the CHANGELOG.md file.
- [ ] LANL employees: make sure tests pass both on the github CI and on the Darwin CI

If preparing for a new release, in addition please check the following:
- [ ] Update the version in cmake.
- [ ] Move the changes in the CHANGELOG.md file under a new header for the new release, and reset the categories.

### CODE REVIEW

File: singularity-eos/base/variadic_utils.hpp
Comment (Yurlungur) [inline singularity-eos/base/variadic_utils.hpp:39]: This can be cleaned up when we move to C++17. Thanks to @dholladay00 for the particular version of this, which checks edge cases.

**Comment by Yurlungur** [comment]
@rbberger please help with the pybind here. We need to force the python module to specialize the lambdas to `Real*`

**Comment by rbberger** [comment]
@Yurlungur turned out to be a trivial change.

**Comment by Yurlungur** [comment]
This is now ready for review. @dholladay00 @gopsub @jhp-lanl

File: singularity-eos/base/variadic_utils.hpp
Comment (jhp-lanl) [reply singularity-eos/base/variadic_utils.hpp:39]: It's too late on a Friday for understanding the nestedness going on with `::type>::type>::value>::type>` 🤯

Comment (Yurlungur) [reply singularity-eos/base/variadic_utils.hpp:39]: Haha blame/thank @dholladay00 for that one. Conceptually all it's doing is checking whether or not T is a pointer. If T is a pointer, check if it's null. Otherwise, return false.

All the other nested type nonsense is checking for edge cases, references to pointers, const pointers, etc.

Comment (Yurlungur) [reply singularity-eos/base/variadic_utils.hpp:39]: The C++17 version will need to retain the nested templates, but will be able to move it out of `std::enable_if` and into `if constexpr`

Comment (jhp-lanl) [reply singularity-eos/base/variadic_utils.hpp:39]: This is something we did in EAP to handle this. It might make sense to implement this to just make things slightly more readable and reusable if we need to do this elsewhere:
```c++
// C++14 implementation of std::remove_cvref (availalbe since C++20)
template<typename T>
struct remove_cvref { typedef std::remove_cv_t<std::remove_reference_t<T>> type; };

// Helper types
template<typename T>
using remove_cvref_t = typename remove_cvref<T>::type;
```

Credit to Diego via CJ for our implementation

Comment (Yurlungur) [reply singularity-eos/base/variadic_utils.hpp]: :+1: Done.

**Comment by Yurlungur** [comment]
Ping @jhp-lanl @dholladay00 reminder to review.

**Review by jhp-lanl — APPROVED** [review]
Looks good to me. I appreciate the `is_nullptr` function. I think it's _very_ clear now what's going on and if I want to wade into the template nonsense I can look at the `variadic_utils.hpp` file 🙂

File: singularity-eos/eos/eos_spiner.hpp
Comment (dholladay00) [inline singularity-eos/eos/eos_spiner.hpp:157]: should all Eos's get these static casts for the default nullptr?

File: singularity-eos/closure/mixed_cell_models.hpp
Comment (Yurlungur) [reply singularity-eos/closure/mixed_cell_models.hpp:569]: Oops good catch. This line is commented out so it doesn't actually matter. But updated.

File: singularity-eos/eos/eos_spiner.hpp
Comment (Yurlungur) [reply singularity-eos/eos/eos_spiner.hpp:157]: It only really matters for the ones I did it for. But I'll change it everywhere for consistency.

**Comment by Yurlungur** [comment]
@dholladay00 I think I addressed all your comments. Can you re-review and approve if you're happy?

**Comment by dholladay00** [comment]
> @dholladay00 I think I addressed all your comments. Can you re-review and approve if you're happy?

Is this passing gitlab CI @Yurlungur ?

**Comment by Yurlungur** [comment]
> > @dholladay00 I think I addressed all your comments. Can you re-review and approve if you're happy?
> 
> Is this passing gitlab CI @Yurlungur ?

Running. Won't merge until it passes.

**Review by dholladay00 — APPROVED** [review]

File: python/module.hpp
Comment (mauneyc-LANL) [inline python/module.hpp:37]: Can this be simplified?
```c++
// somewhere above in scope
constexpr auto realnp(){ return static_cast<Real*>(nullptr); }
// then e.g. 
return (self.*Func)(a, b, realnp());
```
Style preference here, so if not wanted then it's fine as is.

**Review by mauneyc-LANL — APPROVED** [review]

File: singularity-eos/base/variadic_utils.hpp
Comment (Yurlungur) [reply singularity-eos/base/variadic_utils.hpp:41]: Good catch---they actually need to be for device.

File: python/module.hpp
Comment (Yurlungur) [reply python/module.hpp:37]: Good suggestion :+1: I'll switch this at least in the explicit calls.

**Comment by Yurlungur** [comment]
Addressed @mauneyc-LANL 's comments. As soon as tests pass on re-git I will merge.

**Comment by Yurlungur** [comment]
Commenting here for posterity that when we switch to C++17 we will be able to do the following, which will allow us to properly handle passing in `nullptr`:

```C++
using Real = double;

template <typename T>
struct remove_cvref {
  typedef std::remove_cv_t<std::remove_reference_t<T>> type;
};
template <typename T>
using remove_cvref_t = typename remove_cvref<T>::type;

template<typename T>
constexpr inline bool is_null(T &&t) {
    if constexpr(std::is_pointer<remove_cvref_t<T>>::value) {
        return std::forward<T>(t) == nullptr;
    } else {
        return false;
    }
}

#define SG_DO_IF_NOT_NULL(lambda) \
  if constexpr (!std::is_null_pointer<remove_cvref_t<decltype(lambda)>>::value) \
    if (!is_null(lambda)) 

template<typename T>
void DoSomething(T &&t) {
    SG_DO_IF_NOT_NULL(t) {
        std:: cout << t[0] << std::endl;
    }
}

int main() {

    auto q = nullptr;
    std::vector<Real> v = {1, 2, 3, 4};
    DoSomething(v);
    DoSomething(q);

    return 0;
}
```

**Comment by Yurlungur** [comment]
Tests pass on re-git. Merging.

**Pull request merged**
